### ISSUE TITLE
Erratic latency spikes when other users connect to and disconnect from the SoftEtherVPN server

### ISSUE THREAD

**Issue by Alchiadus** [SoftEtherVPN/SoftEtherVPN#949]
### Prerequisites

* [X] Can you reproduce?
* [X] Are you running the latest version of SoftEtherVPN?

**SoftEther version: 24f426162cbbb55ee0fab40a8e3288ea443a3365 (Latest Master branch as of Jul 19, 2019)**
**Component: Server**
**Operating system: Linux (VM)**
**Architecture: 64-bit**
**Processor: Intel® Core™ i7-8700B (Mac mini 2018)**

### Description

I'm experiencing heavy latency spikes when (OpenVPN) users connect to and disconnect from the SoftEtherVPN server. CPU usage increases a little bit during connects and disconnects as can be expected, but not enough to explain spikes I am seeing, nor is the CPU usage close to becoming a bottleneck.

**Expected behavior:**
Ping times remain stable and latency is not (significantly) influenced by other users connecting and disconnecting.

**Actual behavior:**
Ping times spike considerably, sometimes taking as much as several seconds until all new connections' initialisation sequences have completed.

### Steps to reproduce

1. Prepare the SoftEtherVPN server with at least one OpenVPN client configured (tun device, certificate authentication).
2. Connect 50 clients and let them all ping the hub.
3. Observe that ping times spike until all 50 clients are connected.
4. Observe that ping times stabilise once the initialisation sequences have completed.
5. Connect 10 new clients.
6. Observe how ping times spike until all 10 new clients are connected.
7. Observe that ping times stabilise once the initialisation sequences have completed.

I tried the following to see if it had any effect on the erratic ping times:
- Disabling all access rules has no effect
- Setting up multiple users, for example Alice and Bob, and monitor Alice's ping times while many Bobs are connecting has no effect
- Pinging another client instead of the hub has no effect

---

Are there any locks that prevent packets from flowing freely and force them to wait when users connect and disconnect from the server?

---

### Logs

The following two log excerpts are taken from a client pinging the server's hub, in a setting where roughly 50 other users are connected. They demonstrate latency spikes in a scenario where multiple new clients connect at the same time (larger spikes) and in a scenario where only a single new client connects (small spike).

Multiple users connect at roughly the same time, starting from `icmp_seq=2033` to `icmp_seq=2043`:

```
64 bytes from 10.187.0.1: icmp_seq=2020 ttl=128 time=1.32 ms
64 bytes from 10.187.0.1: icmp_seq=2021 ttl=128 time=1.08 ms
64 bytes from 10.187.0.1: icmp_seq=2022 ttl=128 time=1.19 ms
64 bytes from 10.187.0.1: icmp_seq=2023 ttl=128 time=1.43 ms
64 bytes from 10.187.0.1: icmp_seq=2024 ttl=128 time=2.32 ms
64 bytes from 10.187.0.1: icmp_seq=2025 ttl=128 time=1.44 ms
64 bytes from 10.187.0.1: icmp_seq=2026 ttl=128 time=1.26 ms
64 bytes from 10.187.0.1: icmp_seq=2027 ttl=128 time=1.67 ms
64 bytes from 10.187.0.1: icmp_seq=2028 ttl=128 time=1.41 ms
64 bytes from 10.187.0.1: icmp_seq=2029 ttl=128 time=1.89 ms
64 bytes from 10.187.0.1: icmp_seq=2030 ttl=128 time=1.58 ms
64 bytes from 10.187.0.1: icmp_seq=2031 ttl=128 time=1.93 ms
64 bytes from 10.187.0.1: icmp_seq=2032 ttl=128 time=2.28 ms
64 bytes from 10.187.0.1: icmp_seq=2033 ttl=128 time=8.12 ms
64 bytes from 10.187.0.1: icmp_seq=2034 ttl=128 time=229 ms
64 bytes from 10.187.0.1: icmp_seq=2035 ttl=128 time=272 ms
64 bytes from 10.187.0.1: icmp_seq=2036 ttl=128 time=220 ms
64 bytes from 10.187.0.1: icmp_seq=2037 ttl=128 time=421 ms
64 bytes from 10.187.0.1: icmp_seq=2038 ttl=128 time=1450 ms
64 bytes from 10.187.0.1: icmp_seq=2039 ttl=128 time=867 ms
64 bytes from 10.187.0.1: icmp_seq=2040 ttl=128 time=1.32 ms
64 bytes from 10.187.0.1: icmp_seq=2041 ttl=128 time=0.850 ms
64 bytes from 10.187.0.1: icmp_seq=2042 ttl=128 time=835 ms
64 bytes from 10.187.0.1: icmp_seq=2043 ttl=128 time=27.3 ms
64 bytes from 10.187.0.1: icmp_seq=2044 ttl=128 time=1.52 ms
64 bytes from 10.187.0.1: icmp_seq=2045 ttl=128 time=1.24 ms
64 bytes from 10.187.0.1: icmp_seq=2046 ttl=128 time=1.21 ms
64 bytes from 10.187.0.1: icmp_seq=2047 ttl=128 time=2.06 ms
64 bytes from 10.187.0.1: icmp_seq=2048 ttl=128 time=2.03 ms
64 bytes from 10.187.0.1: icmp_seq=2049 ttl=128 time=1.55 ms
64 bytes from 10.187.0.1: icmp_seq=2050 ttl=128 time=2.60 ms
64 bytes from 10.187.0.1: icmp_seq=2051 ttl=128 time=1.13 ms
64 bytes from 10.187.0.1: icmp_seq=2052 ttl=128 time=1.99 ms
64 bytes from 10.187.0.1: icmp_seq=2053 ttl=128 time=2.29 ms
64 bytes from 10.187.0.1: icmp_seq=2054 ttl=128 time=1.62 ms
64 bytes from 10.187.0.1: icmp_seq=2055 ttl=128 time=1.58 ms
64 bytes from 10.187.0.1: icmp_seq=2056 ttl=128 time=2.28 ms
64 bytes from 10.187.0.1: icmp_seq=2057 ttl=128 time=1.39 ms
64 bytes from 10.187.0.1: icmp_seq=2058 ttl=128 time=1.80 ms
64 bytes from 10.187.0.1: icmp_seq=2059 ttl=128 time=1.96 ms
64 bytes from 10.187.0.1: icmp_seq=2060 ttl=128 time=2.55 ms
```

A single user connects, around `icmp_seq=2116` and `icmp_seq=2117`:

```
64 bytes from 10.187.0.1: icmp_seq=2105 ttl=128 time=0.881 ms
64 bytes from 10.187.0.1: icmp_seq=2106 ttl=128 time=1.75 ms
64 bytes from 10.187.0.1: icmp_seq=2107 ttl=128 time=1.90 ms
64 bytes from 10.187.0.1: icmp_seq=2108 ttl=128 time=1.31 ms
64 bytes from 10.187.0.1: icmp_seq=2109 ttl=128 time=1.54 ms
64 bytes from 10.187.0.1: icmp_seq=2110 ttl=128 time=2.03 ms
64 bytes from 10.187.0.1: icmp_seq=2111 ttl=128 time=1.53 ms
64 bytes from 10.187.0.1: icmp_seq=2112 ttl=128 time=1.67 ms
64 bytes from 10.187.0.1: icmp_seq=2113 ttl=128 time=0.848 ms
64 bytes from 10.187.0.1: icmp_seq=2114 ttl=128 time=1.76 ms
64 bytes from 10.187.0.1: icmp_seq=2115 ttl=128 time=2.05 ms
64 bytes from 10.187.0.1: icmp_seq=2116 ttl=128 time=72.3 ms
64 bytes from 10.187.0.1: icmp_seq=2117 ttl=128 time=48.4 ms
64 bytes from 10.187.0.1: icmp_seq=2118 ttl=128 time=2.11 ms
64 bytes from 10.187.0.1: icmp_seq=2119 ttl=128 time=1.80 ms
64 bytes from 10.187.0.1: icmp_seq=2120 ttl=128 time=1.54 ms
64 bytes from 10.187.0.1: icmp_seq=2121 ttl=128 time=1.79 ms
64 bytes from 10.187.0.1: icmp_seq=2122 ttl=128 time=2.09 ms
64 bytes from 10.187.0.1: icmp_seq=2123 ttl=128 time=1.90 ms
64 bytes from 10.187.0.1: icmp_seq=2124 ttl=128 time=1.42 ms
64 bytes from 10.187.0.1: icmp_seq=2125 ttl=128 time=1.67 ms
```

**Comment by dnobori** [comment]
Could you please reproduce the same problem with the stable version of SoftEther VPN?

**Comment by Alchiadus** [comment]
Sure, I will when I get into the office on Monday, I don't have the necessary tools available at the moment.

**Comment by Alchiadus** [comment]
Instead of running everything (server + all clients) on the same machine, I moved the SoftEtherVPN server to a different one. I can still reproduce spikes when 50 users connect at the same time on all of the following versions (I did not test other versions):
- `5.01.9671`
- `5.01.9670`
- `5.01.9669`
- `v4.29-9680-rtm`

I no longer believe that my initial assumption about CPU not being the bottleneck holds as `top`'s default update delay of 3 seconds is not low enough to show them. Decreasing `top`'s update delay to 0.2 seconds shows very brief 100% CPU peaks on local machines.

I repeated the tests by running SoftEtherVPN server on a beefy VPS and can often connect 100 users at the same time without latency spikes. Only sometimes do I observe the spikes I mentioned in my report above and they correlate with very high CPU usage on the server, so I believe it's safe to say it's simply a resource problem.
